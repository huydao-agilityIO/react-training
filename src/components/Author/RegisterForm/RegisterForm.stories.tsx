// RegisterForm.stories.ts|tsx

import type { Meta, StoryObj } from '@storybook/react';
import { MemoryRouter } from 'react-router-dom';
import { FormProvider, useForm } from 'react-hook-form';

// Components
import RegisterForm from './index';

const meta: Meta<typeof RegisterForm> = {
  title: 'Components/Author/RegisterForm',
  component: RegisterForm,
  parameters: {
    layout: 'centered'
  },
  decorators: [
    (Story) => (
      <MemoryRouter>
        <FormProvider {...useForm()}>
          <Story />
        </FormProvider>
      </MemoryRouter>
    )
  ],
  tags: ['autodocs']
};

export default meta;
type Story = StoryObj<typeof RegisterForm>;

export const Default: Story = {
  args: {
    onSubmitForm: () => alert('handles submit form register')
  }
};
