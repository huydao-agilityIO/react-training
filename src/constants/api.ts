export const METHOD = {
  GET: 'GET',
  POST: 'POST',
  PUT: 'PUT',
  DELETE: 'DELETE'
};

export const ERROR_MESSAGES_SERVICES = {
  API: 'API not found!',
  END_POINT: 'Something went wrong!!!'
};

export const END_POINTS = {
  SIGN_IN: 'sign-in',
  SIGN_UP: 'sign-up',
  PATIENT: 'patient',
  HOSPITAL_STAFF: 'hospital-staff'
};

// TODO: Update later
export const API_HOSPITAL_MANAGEMENT = {
  AUTHENTICATION_SIGN_IN: `${process.env.VITE_NX_AUTHENTICATION}/${END_POINTS.SIGN_IN}`,
  AUTHENTICATION_SIGN_UP: `${process.env.VITE_NX_AUTHENTICATION}/${END_POINTS.SIGN_UP}`
};
