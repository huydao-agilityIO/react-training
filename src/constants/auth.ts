export const INVALID_INFO = {
  name: 'Lionel Messi',
  email: 'lionel_messi@co',
  password: 'lio1968'
};

export const VALID_INFO = {
  name: 'Lionel Messi',
  email: 'lionel_messi@gmail.com',
  password: 'lionel_messi1968'
};
