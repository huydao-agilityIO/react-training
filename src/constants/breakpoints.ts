export const breakpoints: Record<string, string> = {
  sm: '414px',
  md: '768px',
  lg: '1024px',
  xl: '1200px',
  '2xl': '1440px'
};
