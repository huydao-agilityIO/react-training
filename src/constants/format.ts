export const CURRENCY_DEFAULT = {
  STYLE: 'currency',
  TEXT: 'USD',
  DECIMAL: 0
};

export const FORMAT_DATE = 'DD/MM/YYYY';
