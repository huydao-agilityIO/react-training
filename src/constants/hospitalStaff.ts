import { HospitalStaff } from '@shared/types';

export const HOSPITAL_STAFF: HospitalStaff[] = [
  {
    firstName: 'Jovan',
    lastName: 'Crist',
    urlAvatar:
      'https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/931.jpg',
    description: 'description 1',
    id: '1'
  },
  {
    firstName: 'Lazaro',
    lastName: 'Schumm',
    urlAvatar:
      'https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/158.jpg',
    description: 'description 2',
    id: '2'
  },
  {
    firstName: 'Vidal',
    lastName: 'Russel',
    urlAvatar:
      'https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/1158.jpg',
    description: 'description 3',
    id: '3'
  },
  {
    firstName: 'Jed',
    lastName: 'Kovacek',
    urlAvatar:
      'https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/1085.jpg',
    description: 'description 4',
    id: '4'
  },
  {
    firstName: 'Tatyana',
    lastName: 'Koepp',
    urlAvatar:
      'https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/1144.jpg',
    description: 'description 5',
    id: '5'
  }
];
