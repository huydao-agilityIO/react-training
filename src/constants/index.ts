export * from './api';
export * from './auth';
export * from './format';
export * from './patient';
export * from './route';
export * from './sidebar';
export * from './validate';
