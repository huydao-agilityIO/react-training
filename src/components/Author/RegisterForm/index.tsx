import { FormEventHandler } from 'react';
import { Control, Controller } from 'react-hook-form';
import { Button, Checkbox, HStack, Spinner } from '@chakra-ui/react';

// Utils
import { FORM_SCHEMA } from '@shared/utils';

// Types
import { Authentication } from '@shared/types';

// Constants
import { ROUTE } from '@shared/constants';

// Components
import AuthFormContainer from '../AuthFormContainer';
import { FormGroupControl } from '@shared/components';

interface AuthFormProps {
  control?: Control<Authentication>;
  isLoading?: boolean;
  errorResponseAPI?: string;
  onSubmitForm?: FormEventHandler<HTMLFormElement>;
}

const RegisterForm = ({
  isLoading = false,
  errorResponseAPI = '',
  control,
  onSubmitForm
}: AuthFormProps) => (
  <AuthFormContainer
    idForm="registerFom"
    mainTitle="Register"
    linkRouter="Log in"
    pathRouter={ROUTE.LOGIN_PAGE}
    errorResponseAPI={errorResponseAPI}
    onSubmit={onSubmitForm}>
    <HStack w="full">
      <Controller
        name="firstName"
        control={control}
        rules={FORM_SCHEMA.FIRST_NAME}
        render={({
          field: { value, onChange, ...rest },
          fieldState: { error }
        }) => {
          return (
            <FormGroupControl
              type="text"
              id="firstName"
              placeholder="First Name"
              value={value}
              errorMessage={error?.message}
              onChange={onChange}
              {...rest}
            />
          );
        }}
      />
      <Controller
        name="lastName"
        control={control}
        rules={FORM_SCHEMA.LAST_NAME}
        render={({
          field: { value, onChange, ...rest },
          fieldState: { error }
        }) => {
          return (
            <FormGroupControl
              type="text"
              id="lastName"
              placeholder="Last Name"
              value={value}
              errorMessage={error?.message}
              onChange={onChange}
              {...rest}
            />
          );
        }}
      />
    </HStack>
    <Controller
      name="email"
      control={control}
      rules={FORM_SCHEMA.EMAIL}
      render={({
        field: { value, onChange, ...rest },
        fieldState: { error }
      }) => {
        return (
          <FormGroupControl
            type="email"
            id="email"
            placeholder="Enter your email"
            value={value}
            errorMessage={error?.message}
            onChange={onChange}
            {...rest}
          />
        );
      }}
    />
    <Controller
      name="password"
      control={control}
      rules={FORM_SCHEMA.PASSWORD}
      render={({
        field: { value, onChange, ...rest },
        fieldState: { error }
      }) => {
        return (
          <FormGroupControl
            type="password"
            id="password"
            placeholder="Password"
            value={value}
            errorMessage={error?.message}
            onChange={onChange}
            {...rest}
          />
        );
      }}
    />
    <Checkbox
      size="lg"
      color="dark.50"
      colorScheme="gray"
      isDisabled={isLoading}>
      Keep me up to date
    </Checkbox>

    <Button
      type="submit"
      w="full"
      isLoading={isLoading}
      spinner={<Spinner color="light.300" />}>
      Sign up
    </Button>
  </AuthFormContainer>
);

export default RegisterForm;
