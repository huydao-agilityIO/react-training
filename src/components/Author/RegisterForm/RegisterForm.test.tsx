import { BrowserRouter } from 'react-router-dom';
import { FormProvider, useForm } from 'react-hook-form';

// Constants
import { VALID_INFO } from '@shared/constants';

// Components
import RegisterForm from '.';

describe('RegisterForm', () => {
  const mockSubmitForm = jest.fn();
  const defaultValues = {
    fullName: '',
    email: '',
    password: ''
  };
  const { result } = testLibReactUtils.renderHook(() =>
    useForm({ defaultValues })
  );
  const ContainerRegisterForm = () => (
    <BrowserRouter>
      <FormProvider {...result.current}>
        <RegisterForm onSubmitForm={mockSubmitForm} />
      </FormProvider>
    </BrowserRouter>
  );

  it('renders correctly', () => {
    const { container } = testLibReactUtils.render(<ContainerRegisterForm />);

    expect(container).toMatchSnapshot();
  });

  it('submits the form with correct values', () => {
    const { getByPlaceholderText, getByText } = testLibReactUtils.render(
      <ContainerRegisterForm />
    );

    testLibReactUtils.fireEvent.change(getByPlaceholderText('First Name'), {
      target: { value: VALID_INFO.name }
    });
    testLibReactUtils.fireEvent.change(
      getByPlaceholderText('Enter your email'),
      {
        target: { value: VALID_INFO.email }
      }
    );
    testLibReactUtils.fireEvent.change(getByPlaceholderText('Password'), {
      target: { value: VALID_INFO.password }
    });
    testLibReactUtils.fireEvent.submit(getByText('Sign up'));

    expect(mockSubmitForm).toHaveBeenCalled();
  });
});
