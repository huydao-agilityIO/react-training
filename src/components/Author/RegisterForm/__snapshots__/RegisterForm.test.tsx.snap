// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterForm renders correctly 1`] = `
<div>
  <div
    class="chakra-stack css-1hg7tnx"
  >
    <div
      class="chakra-container css-1s7vqq3"
    >
      <form
        class="css-0"
        id="registerFom"
      >
        <div
          class="chakra-stack css-1e1otyl"
        >
          <div
            class="chakra-stack css-rgfwhv"
          >
            <h2
              class="chakra-heading css-tro5bq"
            >
              Register
            </h2>
          </div>
          <div
            class="chakra-stack css-feqp41"
          >
            <div
              class="chakra-stack css-9kwbnc"
            >
              <div
                class="chakra-stack css-10t90fk"
              >
                <div
                  class="chakra-form-control css-1kxonj9"
                  role="group"
                >
                  <input
                    class="chakra-input css-16v8qx8"
                    id="firstName"
                    placeholder="First Name"
                    type="text"
                    value=""
                  />
                </div>
                <div
                  class="chakra-form-control css-1kxonj9"
                  role="group"
                >
                  <input
                    class="chakra-input css-16v8qx8"
                    id="lastName"
                    placeholder="Last Name"
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                class="chakra-form-control css-1kxonj9"
                role="group"
              >
                <input
                  class="chakra-input css-16v8qx8"
                  id="email"
                  placeholder="Enter your email"
                  type="email"
                  value=""
                />
              </div>
              <div
                class="chakra-form-control css-1kxonj9"
                role="group"
              >
                <input
                  class="chakra-input css-16v8qx8"
                  id="password"
                  placeholder="Password"
                  type="password"
                  value=""
                />
              </div>
              <label
                class="chakra-checkbox css-162ix9b"
              >
                <input
                  class="chakra-checkbox__input"
                  style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                  type="checkbox"
                  value=""
                />
                <span
                  aria-hidden="true"
                  class="chakra-checkbox__control css-ca8wb8"
                />
                <span
                  class="chakra-checkbox__label css-r90m2q"
                >
                  Keep me up to date
                </span>
              </label>
              <button
                class="chakra-button css-1lsg31x"
                type="submit"
              >
                Sign up
              </button>
            </div>
            <div
              class="chakra-stack css-12yeauu"
            >
              <p
                class="chakra-text css-oi2y7z"
              >
                Need an account?
                 
                <a
                  class="chakra-link css-jqihjn"
                  href="/"
                >
                  Log in
                </a>
              </p>
              <a
                class="chakra-link css-3ctcoe"
                href="/"
              >
                Forget Password?
              </a>
            </div>
          </div>
        </div>
      </form>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
